src/Features/Core/Portable/QuickInfo/CommonSemanticQuickInfoProvider.cs

```diff
@@ -186,9 +188,80 @@ protected static Task<QuickInfoItem> CreateContentAsync(
+
+            // Look to see if we're inside a suppression (e.g. `expr!`).  The suppression changes the nullability analysis,
+            // and we don't actually want that here as we want to show the original nullability prior to the suppression applying.
+            //
+            // In that case, actually fork the semantic model with the `!` removed and then re-bind the token, getting the 
+            // analysis results from that.
+            var tokenParent = token.GetRequiredParent();
+            var parentSuppression = GetOuterSuppression(tokenParent);
+            if (parentSuppression is null)
+                return false;
+
+            var root = semanticModel.SyntaxTree.GetRoot(cancellationToken);
+
+            var editor = new SyntaxEditor(root, services);
+            // First, mark the token, so we can find it later.
+            editor.ReplaceNode(
+                tokenParent, tokenParent.ReplaceToken(token, token.WithAdditionalAnnotations(s_annotation)));
+
+            // Now walk upwards, removing all the suppressions until we hit the top of the suppression chain.
+            for (var currentSuppression = parentSuppression;
+                 currentSuppression is not null;
+                 currentSuppression = GetOuterSuppression(currentSuppression))
+            {
+                editor.ReplaceNode(
+                    currentSuppression,
+                    (current, _) => syntaxFacts.GetOperandOfPostfixUnaryExpression(current));
+            }
+
+            // Now fork the semantic model with the new root that has the suppressions removed.
+            var newRoot = editor.GetChangedRoot();
+
+            var newTree = semanticModel.SyntaxTree.WithRootAndOptions(newRoot, semanticModel.SyntaxTree.Options);
+            var newToken = newTree.GetRoot(cancellationToken).GetAnnotatedTokens(s_annotation).Single();
+
+            var newBindableParent = syntaxFacts.TryGetBindableParent(newToken);
+            if (newBindableParent is null)
+                return false;
+
+            var newCompilation = semanticModel.Compilation.ReplaceSyntaxTree(semanticModel.SyntaxTree, newTree);
+            semanticModel = newCompilation.GetSemanticModel(newTree);
+
+            var symbols = BindSymbols(services, semanticModel, newToken, cancellationToken);
+            if (symbols.IsEmpty)
+                return false;
+
+            analysis = GetNullabilityAnalysis(semanticModel, symbols[0], newBindableParent, cancellationToken);
+            return true;
+
+            SyntaxNode? GetOuterSuppression(SyntaxNode node)
+                => node.Ancestors().FirstOrDefault(a => a.RawKind == syntaxFacts.SyntaxKinds.SuppressNullableWarningExpression);
+        }
+    }
+
+    protected virtual (NullableAnnotation, NullableFlowState) GetNullabilityAnalysis(SemanticModel semanticModel, ISymbol symbol, SyntaxNode node, CancellationToken cancellationToken)
+        => default;
+
+    protected ImmutableArray<ISymbol> BindSymbols(
```

Comment:
Broke BindToken into BindSymbols (just tries to get teh symbols corresponding to a token).  That way that helper can be used when we decide to rebind speculatively when inside a suppression